### PR TITLE
[server] try to fix org membership on login

### PR DESCRIPTION
**Pull request by AlexTugarev**
in case it didn't succeed on first attempt this might help during the following ones.

## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 69f54c6</samp>

Refactor `iam-session-app.ts` to optimize login with OIDC and organization membership. Reduce database queries and check organization validity.

</details>

## Related Issue(s)

Fixes #

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [x] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/iam/iam-session-app.ts [hunk 1]
```diff
@@ -98,17 +98,31 @@ export class IamSessionApp {
...
-            log.info("Found Org-owned user by email.", { email: payload?.claims?.email });
+        if (existingUser?.organizationId) {
+            const members = await this.orgService.listMembers(existingUser.id, existingUser.organizationId);
+            if (!members.some((m) => m.userId === existingUser?.id)) {
+                // In case `createNewOIDCUser` failed to create a membership for this user,
+                // let's try to fix the situation on the fly.
+                // Also, if that step repeatedly fails, it would fail the login process earlier but
+                // in a more consistent state.
+                await this.orgService.addOrUpdateMember(
+                    existingUser.id,
```
Comment (svenefftinge) [inline components/server/src/iam/iam-session-app.ts:120]: that user would not have the permission to add themselves. We need to allow `addOrUpdateMember` without subject for the iam (ie.. system) use case.

**Review by svenefftinge — APPROVED** [review]
LGTM to me. I can handle my comment in a follow up PR as I need to change this in the other usage as well.

File: components/server/src/iam/iam-session-app.ts (hunk 1 above)
Comment (svenefftinge) [reply components/server/src/iam/iam-session-app.ts]: added a commit to solve this.

File: components/server/src/orgs/organization-service.ts [hunk 2]
```diff
@@ -179,12 +179,14 @@ export class OrganizationService {
...
 
     public async addOrUpdateMember(
-        userId: string,
+        userId: string | undefined, // undefined means it is a system call, not a user call
         orgId: string,
         memberId: string,
         role: OrgMemberRole,
     ): Promise<void> {
-        await this.auth.checkPermissionOnOrganization(userId, "write_members", orgId);
+        if (userId) {
```
Comment (AlexTugarev) [inline components/server/src/orgs/organization-service.ts:187]: @svenefftinge, now thinking about that change. This could actually be the fix.

Comment (AlexTugarev) [reply components/server/src/orgs/organization-service.ts:187]: Just curious, why this is not happening on dogfood?

Comment (svenefftinge) [reply components/server/src/orgs/organization-service.ts:187]: The featureflag is disabled. So these checks don't throw

**Pull request merged**